patroni/dcs/__init__.py

```diff
@@ -1131,50 +1146,67 @@ def _get_members_slots(self, name: str, role: str) -> Dict[str, Dict[str, str]]:
         # also exlude members with disabled WAL streaming
         members = filter(lambda m: m.name != name and not m.nostream, self.members)
 
-        if role in ('master', 'primary', 'standby_leader'):
+        if can_advance_slots and global_config.member_slots_ttl > 0:
+            # if the node does only cascading and can't become the leader, we
+            # want only to have slots for members that could connect to it.
+            members = [m for m in members if not nofailover or m.replicatefrom == name]
+        elif role in ('master', 'primary', 'standby_leader'):  # PostgreSQL is older than 11
+            # on the leader want to have slots only for the nodes that are supposed to be replicating from it.
```

Comment:
> I was wondering about that because the primary can also fall into the current first if (about slots and ttl), can't it?

This is exactly the reason.
Normally (old behavior), primary should have only replication slots for nodes that are streaming from it (exclude nodes with `tags.replicatefrom: othernode` if the `othernode` is healthy):
```
primary<-othernode<-node2
```

But, with the new feature it needs to be different:
```
primary: slots(othernode, node2)
othernode: slots(primary, node2)
node2: slots(primary, othernode)
```

That is, `primary` will have active slot for `othernode` and inactive slot for `node2`, which needs periodically to move forward using `pg_replication_slot_advance()`.
The same story for remaining nodes.

Nodes with `tags.nofailover` that are doing cascading are different. They need to have slots only for nodes that are supposed to replicate from them (`if not nofailover or m.replicatefrom == name`).

Reply:
Got it, thanks for clarifying!